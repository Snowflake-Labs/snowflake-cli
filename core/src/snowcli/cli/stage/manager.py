from __future__ import annotations

from pathlib import Path

from snowcli.cli.common.sql_execution import SqlExecutionMixin


class StageManager(SqlExecutionMixin):
    @staticmethod
    def get_standard_stage_name(name: str) -> str:
        # Handle embedded stages
        if name.startswith("snow://"):
            return name

        return f"@{name}"

    def list(self, stage_name: str):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"ls {stage_name}")

    def get(self, stage_name: str, dest_path: Path):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"get {stage_name} file://{dest_path}/")

    def put(self, local_path: str, stage_name: str, parallel: int, overwrite: bool):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(
            f"put file://{local_path} {stage_name} "
            f"auto_compress=false parallel={parallel} overwrite={overwrite}"
        )

    def remove(self, stage_name: str, path: str):
        stage_name = self.get_standard_stage_name(stage_name)
        path = path if path.startswith("/") else "/" + path
        return self._execute_query(f"remove {stage_name}{path}")

    def show(self):
        return self._execute_query("show stages")

    def create(self, stage_name: str):
        return self._execute_query(f"create stage if not exists {stage_name}")

    def drop(self, stage_name: str):
        return self._execute_query(f"drop stage {stage_name}")
