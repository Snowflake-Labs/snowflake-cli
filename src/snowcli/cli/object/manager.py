from __future__ import annotations

from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowflake.connector.cursor import SnowflakeCursor


class ObjectManager(SqlExecutionMixin):
    def show(self, object_type: str, like: str) -> SnowflakeCursor:
        return self._execute_query(f"show {object_type} like '{like}'")

    def drop(self, object_type, name: str) -> SnowflakeCursor:
        return self._execute_query(f"drop {object_type} {name}")

    def describe(self, object_type: str, name: str):
        return self._execute_query(f"describe {object_type} {name}")
