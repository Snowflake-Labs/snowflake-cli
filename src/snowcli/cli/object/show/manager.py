from __future__ import annotations

from snowflake.connector.cursor import SnowflakeCursor

from snowcli.cli.common.sql_execution import SqlExecutionMixin


class ShowManager(SqlExecutionMixin):
    def show(self) -> SnowflakeCursor:
        return self._execute_query("show warehouses")
