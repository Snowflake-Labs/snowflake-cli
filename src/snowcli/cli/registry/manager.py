from snowcli.cli.common.flags import ConnectionOption
from snowcli.cli.common.snow_cli_global_context import snow_cli_global_context_manager
from snowcli.snow_connector import connect_to_snowflake


def get_token():
    """
    Get token to authenticate with registry.
    """
    conn = snow_cli_global_context_manager.get_connection()

    # disable session deletion
    conn._all_async_queries_finished = lambda: False
    if conn._rest is None:
        raise Exception("error in connection object")
    # obtain and create the token
    token_data = conn._rest._token_request("ISSUE")

    return {
        "token": token_data["data"]["sessionToken"],
        "expires_in": token_data["data"]["validityInSecondsST"],
    }
