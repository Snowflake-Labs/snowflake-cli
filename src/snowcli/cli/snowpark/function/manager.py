from __future__ import annotations

from typing import Optional, List

import click

from snowcli.cli.common.sql_execution import SqlExecutionMixin


class FunctionManager(SqlExecutionMixin):
    @staticmethod
    def identifier(
        name: Optional[str] = None,
        signature: Optional[str] = None,
        name_and_signature: Optional[str] = None,
    ):
        if all([name, signature, name_and_signature]):
            raise click.ClickException(
                "Provide only one, name and arguments or full signature. Both provided."
            )

        if not (name and signature) and not name_and_signature:
            raise click.ClickException(
                "Provide either name and arguments or full signature. None provided."
            )

        if name and signature:
            name_and_signature = name + signature

        return name_and_signature

    def drop(self, identifier: str):
        return self._execute_query(f"drop function {identifier}")

    def show(self, like: Optional[str] = None):
        query = "show user functions"
        if like:
            query += f" like '{like}'"
        return self._execute_query(query)

    def describe(self, identifier: str):
        return self._execute_query(f"describe function {identifier}")

    def execute(self, expression: str):
        return self._execute_query(f"select {expression}")

    @staticmethod
    def artifact_stage_path(identifier: str):
        return (
            identifier.replace(
                "(",
                "",
            )
            .replace(
                ")",
                "",
            )
            .replace(
                " ",
                "_",
            )
            .replace(
                ",",
                "",
            )
            .lower()
        )

    def create(
        self,
        identifier: str,
        return_type: str,
        handler: str,
        artifact_file: str,
        packages: List[str],
        overwrite: bool,
    ):
        create_stmt = "create or replace" if overwrite else "create"
        return self._execute_query(
            f"""\
            {create_stmt} function {identifier}
            returns {return_type}
            language python
            runtime_version=3.8
            imports=('{artifact_file}')
            handler='{handler}'
            packages=()
        """
        )
