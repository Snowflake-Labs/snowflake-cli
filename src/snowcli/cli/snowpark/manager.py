from __future__ import annotations

import logging
from typing import List

from snowcli.cli.snowpark.common import SnowparkObjectManager
from snowflake.connector.cursor import SnowflakeCursor

log = logging.getLogger(__file__)


class FunctionManager(SnowparkObjectManager):
    @property
    def _object_type(self):
        return "function"

    @property
    def _object_execute(self):
        return "select"

    def create_or_replace(
        self,
        identifier: str,
        return_type: str,
        handler: str,
        artifact_file: str,
        packages: List[str],
    ) -> SnowflakeCursor:
        log.debug(f"Creating function {identifier} using @{artifact_file}")
        packages_list = ",".join(f"'{p}'" for p in packages)
        return self._execute_query(
            f"""\
            create or replace function {identifier}
            returns {return_type}
            language python
            runtime_version=3.8
            imports=('{artifact_file}')
            handler='{handler}'
            packages=({packages_list})
        """
        )


class ProcedureManager(SnowparkObjectManager):
    @property
    def _object_type(self):
        return "procedure"

    @property
    def _object_execute(self):
        return "call"

    def create_or_replace(
        self,
        identifier: str,
        return_type: str,
        handler: str,
        artifact_file: str,
        packages: List[str],
        execute_as_caller: bool = False,
    ) -> SnowflakeCursor:
        log.debug(f"Creating procedure {identifier} using @{artifact_file}")
        packages_list = ",".join(f"'{p}'" for p in packages)
        query = f"""\
            create or replace procedure {identifier}
            returns {return_type}
            language python
            runtime_version=3.8
            imports=('{artifact_file}')
            handler='{handler}'
            packages=({packages_list})
        """
        if execute_as_caller:
            query += "\nexecute as caller"
        return self._execute_query(query)
