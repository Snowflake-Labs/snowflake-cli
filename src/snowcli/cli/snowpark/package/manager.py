from __future__ import annotations
import logging
import os.path
import tempfile
from pathlib import Path


from requirements.requirement import Requirement
from shutil import rmtree

from snowcli import utils
from snowcli.cli.common.snow_cli_global_context import snow_cli_global_context_manager
from snowcli.cli.snowpark.package.utils import (
    LookupResult,
    InAnaconda,
    RequiresPackages,
    NotInAnaconda,
    NothingFound,
    CreatedSuccessfully,
    CreationError,
)
from snowcli.utils import SplitRequirements

log = logging.getLogger(__name__)


class PackageManager:
    @staticmethod
    def lookup(name: str, install_packages: bool) -> LookupResult:

        package_response = utils.parse_anaconda_packages([Requirement.parse(name)])

        if package_response.snowflake and not package_response.other:
            return InAnaconda(package_response)
        else:
            if install_packages:
                status, result = utils.install_packages(
                    perform_anaconda_check=True, package_name=name, file_name=None
                )

                if status:
                    if result.snowflake:
                        return RequiresPackages(result)
                    else:
                        return NotInAnaconda(result)

        return NothingFound(SplitRequirements([], []))

    def upload(self, file: Path, stage: str, overwrite: bool):
        conn = snow_cli_global_context_manager.get_connection()

        log.info(f"Uploading {file} to Snowflake @{stage}/{file}...")
        with tempfile.TemporaryDirectory() as temp_dir:
            temp_app_zip_path = utils.prepare_app_zip(file, temp_dir)
            deploy_response = conn.upload_file_to_stage(
                file_path=temp_app_zip_path,
                destination_stage=stage,
                path="/",
                database=conn.ctx.database,
                schema=conn.ctx.schema,
                overwrite=overwrite,
                role=conn.ctx.role,
                warehouse=conn.ctx.warehouse,
            )
        log.info(
            f"Package {file} {deploy_response.description[6]} to Snowflake @{stage}/{file}."
        )
        if deploy_response.description[6] == "SKIPPED":
            log.info(
                "Package already exists on stage. Consider using --overwrite to overwrite the file."
            )

    def create(self, name: str):
        file_name = name + ".zip"
        if os.path.exists(".packages"):
            utils.recursive_zip_packages_dir(pack_dir=".packages", dest_zip=file_name)

        if os.path.exists(file_name):
            return CreatedSuccessfully(Path(file_name))
        else:
            return CreationError()

    @staticmethod
    def cleanup_after_install():
        if os.path.exists(".packages"):
            rmtree(".packages")

    @staticmethod
    def create_lookup_message(lookup_result: LookupResult, name: str):

        if type(lookup_result) == InAnaconda:
            return f"Package {name} is available on the Snowflake anaconda channel."
        elif type(lookup_result) == RequiresPackages:
            return f"""The package {name} is supported, but does depend on the
                    following Snowflake supported native libraries. You should
                    include the following in your packages: {lookup_result.requirements.snowflake}"""
        elif type(lookup_result) == NotInAnaconda:
            return f"""The package {name} is avaiable through PIP. You can create a zip using:\n
                    snow snowpark package create {name} -y"""
        else:
            return f"Lookup for package {name} resulted in some error. Please check the package name and try again"
