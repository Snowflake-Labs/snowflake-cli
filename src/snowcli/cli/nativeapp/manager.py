from __future__ import annotations
from collections import OrderedDict

import logging
from pathlib import Path
from functools import cached_property
from typing import List, Optional
from click.exceptions import ClickException
from snowcli.exception import SnowflakeSQLExecutionError

from snowflake.connector.cursor import SnowflakeCursor

from snowcli.cli.project.util import clean_identifier
from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowcli.cli.project.definition import (
    default_app_package,
    default_role,
    default_application,
)
from snowcli.output.printing import print_result
from snowcli.output.types import ObjectResult
from snowcli.cli.stage.diff import (
    DiffResult,
    stage_diff,
    sync_local_diff_with_stage,
)
from snowcli.cli.project.definition_manager import DefinitionManager
from snowcli.cli.nativeapp.artifacts import (
    build_bundle,
    translate_artifact,
    ArtifactMapping,
)

from snowflake.connector.cursor import DictCursor

SPECIAL_COMMENT = "GENERATED_BY_SNOWCLI"
COMMENT_COL = "comment"

log = logging.getLogger(__name__)


class ApplicationPackageAlreadyExistsError(ClickException):
    def __init__(self, name: str):
        super().__init__(
            f"An Application Package {name} already exists in account that may have been created without snowCLI. "
        )


class CouldNotDropObjectError(ClickException):
    """
    Could not successfully drop the required Snowflake object.
    """

    def __init__(self, message: str):
        super().__init__(message=message)


class NativeAppManager(SqlExecutionMixin):
    definition_manager: DefinitionManager

    def __init__(self, search_path: Optional[str] = None):
        super().__init__()
        self.definition_manager = DefinitionManager(search_path or "")

    @property
    def project_root(self) -> Path:
        return self.definition_manager.project_root

    @property
    def definition(self) -> dict:
        return self.definition_manager.project_definition["native_app"]

    @cached_property
    def artifacts(self) -> List[ArtifactMapping]:
        return [translate_artifact(item) for item in self.definition["artifacts"]]

    @cached_property
    def deploy_root(self) -> Path:
        return Path(self.project_root, self.definition["deploy_root"])

    def build_bundle(self) -> None:
        build_bundle(self.project_root, self.deploy_root, self.artifacts)

    def sync_deploy_root_with_stage(self, role: str, app_pkg: str):
        schema_and_stage: str = self.definition["source_stage"]

        # Does a stage already exist within the app pkg, or we need to create one?
        # Using "if not exists" should take care of either case.
        log.info("Checking if stage exists, or creating a new one if none exists.")
        with self.use_role(role):
            self._execute_query(
                f"create schema if not exists {app_pkg}.{schema_and_stage.split('.')[0]}"
            )
            self._execute_query(
                f"""
                    create stage if not exists {app_pkg}.{schema_and_stage}
                    encryption = (TYPE = 'SNOWFLAKE_SSE')"""
            )

        # Perform a diff operation and display results to the user for informational purposes
        log.info(
            f"Performing a diff between the Snowflake stage and your local deploy_root {self.deploy_root} directory."
        )
        diff: DiffResult = stage_diff(self.deploy_root, f"{app_pkg}.{schema_and_stage}")
        log.info("Listing results of diff:")
        log.info(f"New files only on your local: {','.join(diff.only_local)}")
        log.info(f"New files only on the stage: {','.join(diff.only_on_stage)}")
        log.info(
            f"Existing files modified or status unknown: {','.join(diff.different)}"
        )
        log.info(f"Existing files identical to the stage: {','.join(diff.identical)}")

        # Upload diff-ed files to app pkg stage
        sync_local_diff_with_stage(
            role=role,
            deploy_root_path=self.deploy_root,
            diff_result=diff,
            stage_path=f"{app_pkg}.{schema_and_stage}",
        )

    def app_run(self) -> SnowflakeCursor:

        project_name = clean_identifier(self.definition["name"])
        app_pkg = self.definition.get("package", {}).get(
            "name", default_app_package(project_name)
        )
        app_pkg_role = self.definition.get("package", {}).get("role", default_role())

        with self.use_role(app_pkg_role):
            show_cursor = self._execute_query(
                f"show application packages like '{app_pkg}'",
                cursor_class=DictCursor,
            )

            if show_cursor.rowcount is None:
                raise SnowflakeSQLExecutionError()
            elif show_cursor.rowcount == 0:
                # Create an app pkg, with distribution = internal to avoid triggering security scan
                log.info(f"Creating new application package {app_pkg} in account.")
                self._execute_query(
                    f"""
                    create application package {app_pkg}
                        comment = {SPECIAL_COMMENT}
                        distribution = internal
                """
                )
            else:
                # There can only be one possible pre-existing app pkg with the same name
                show_app_row = show_cursor.fetchone()
                row_comment = show_app_row[
                    COMMENT_COL
                ]  # Because we use a DictCursor, we are guaranteed a dictionary instead of indexing through a list.

                if row_comment != SPECIAL_COMMENT:
                    raise ApplicationPackageAlreadyExistsError(app_pkg)

            # Upload files from deploy root local folder to the above stage
            self.sync_deploy_root_with_stage(app_pkg_role, app_pkg)

    def drop_object(
        self,
        object_name: str,
        object_role: str,
        object_type: str,
        query_dict: dict,
    ) -> None:
        log_object_type = (
            "Application Package" if object_type == "package" else object_type
        )

        with self.use_role(object_role):
            show_obj_cursor = self._execute_query(
                f"{query_dict['show']} '{object_name}'", cursor_class=DictCursor
            )

            if show_obj_cursor.rowcount == 0:
                raise CouldNotDropObjectError(
                    f"Role {object_role} does not own any {log_object_type.lower()} with name {object_name}!"
                )
            elif show_obj_cursor.rowcount > 0:
                # There can only be one possible pre-existing app pkg with the same name
                show_obj_row = show_obj_cursor.fetchone()
                row_comment = show_obj_row[
                    COMMENT_COL
                ]  # Because we use a DictCursor, we are guaranteed a dictionary instead of indexing through a list.

                if row_comment != SPECIAL_COMMENT:
                    raise CouldNotDropObjectError(
                        f"{log_object_type} {object_name} was not created by SnowCLI. Cannot drop the {log_object_type.lower()}."
                    )
            else:  # rowcount is None
                raise SnowflakeSQLExecutionError()

            log.info(f"Dropping {log_object_type.lower()} {object_name} now.")
            self._execute_query(f"{query_dict['drop']} {object_name}")
            log.info(f"Dropped {log_object_type.lower()} {object_name} successfully.")

    def teardown(self) -> None:
        project_name = clean_identifier(self.definition["name"])

        # Drop the application first
        self.drop_object(
            object_name=self.definition.get("application", {}).get(
                "name", default_application(project_name)
            ),
            object_role=self.definition.get("application", {}).get(
                "role", default_role(project_name)
            ),
            object_type="application",
            query_dict={"show": "show applications like", "drop": "drop application"},
        )

        # Drop the application package next
        self.drop_object(
            object_name=self.definition.get("package", {}).get(
                "name", default_app_package(project_name)
            ),
            object_role=self.definition.get("package", {}).get(
                "role", default_role(project_name)
            ),
            object_type="package",
            query_dict={
                "show": "show application packages like",
                "drop": "drop application package",
            },
        )
