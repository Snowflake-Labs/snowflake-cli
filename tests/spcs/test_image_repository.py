import snowflake.cli.plugins.spcs.image_repository.manager
from tests.testing_utils.fixtures import *
import json
from snowflake.cli.plugins.spcs.image_repository.manager import ImageRepositoryManager
from typing import Dict
from click import ClickException


MOCK_ROWS = [
    [
        "2023-01-01 00:00:00",
        "IMAGES",
        "DB",
        "SCHEMA",
        "orgname-alias.registry.snowflakecomputing.com/DB/SCHEMA/IMAGES",
        "ROLE",
        "ROLE",
        "",
    ]
]

MOCK_COLUMNS = [
    "created_on",
    "name",
    "database_name",
    "schema_name",
    "repository_url",
    "owner",
    "owner_role_type",
    "comment",
]
MOCK_ROWS_DICT = [
    {col_name: col_val for col_name, col_val in zip(MOCK_COLUMNS, row)}
    for row in MOCK_ROWS
]


@mock.patch("snowflake.cli.plugins.spcs.image_repository.commands.requests.get")
@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.commands.ImageRepositoryManager._execute_query"
)
@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.commands.ImageRepositoryManager._conn"
)
@mock.patch(
    "snowflake.cli.plugins.spcs.image_registry.commands.RegistryManager.login_to_registry"
)
def test_list_images(
    mock_login,
    mock_conn,
    mock_execute,
    mock_get_images,
    runner,
    mock_cursor,
):
    mock_conn.database = "DB"
    mock_conn.schema = "SCHEMA"
    mock_conn.role = "MY_ROLE"

    mock_execute.return_value = mock_cursor(
        rows=MOCK_ROWS_DICT,
        columns=MOCK_COLUMNS,
    )
    mock_login.return_value = "TOKEN"

    mock_get_images.return_value.status_code = 200
    mock_get_images.return_value.text = '{"repositories":["baserepo/super-cool-repo"]}'

    result = runner.invoke(
        ["spcs", "image-repository", "list-images", "IMAGES", "--format", "JSON"]
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [{"image": "/DB/SCHEMA/IMAGES/super-cool-repo"}]


@mock.patch("snowflake.cli.plugins.spcs.image_repository.commands.requests.get")
@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager._execute_query"
)
@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.commands.ImageRepositoryManager._conn"
)
@mock.patch(
    "snowflake.cli.plugins.spcs.image_registry.manager.RegistryManager.login_to_registry"
)
def test_list_tags(
    mock_login,
    mock_conn,
    mock_execute,
    mock_get_tags,
    runner,
    mock_cursor,
):
    mock_conn.database = "DB"
    mock_conn.schema = "SCHEMA"
    mock_conn.role = "MY_ROLE"

    mock_execute.return_value = mock_cursor(rows=MOCK_ROWS_DICT, columns=MOCK_COLUMNS)
    mock_login.return_value = "TOKEN"

    mock_get_tags.return_value.status_code = 200
    mock_get_tags.return_value.text = (
        '{"name":"baserepo/super-cool-repo","tags":["1.2.0"]}'
    )

    result = runner.invoke(
        [
            "spcs",
            "image-repository",
            "list-tags",
            "IMAGES",
            "--image_name",
            "/DB/SCHEMA/IMAGES/super-cool-repo",
            "--format",
            "JSON",
        ]
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {"tag": "/DB/SCHEMA/IMAGES/super-cool-repo:1.2.0"}
    ]


@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager.get_repository_url"
)
def test_get_repository_url_cli(mock_url, runner):
    repo_url = "repotest.registry.snowflakecomputing.com/db/schema/IMAGES"
    mock_url.return_value = repo_url
    result = runner.invoke(["spcs", "image-repository", "url", "IMAGES"])
    assert result.exit_code == 0, result.output
    assert result.output.strip() == repo_url


@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager._execute_schema_query"
)
def test_get_repository_row(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor(
        rows=MOCK_ROWS_DICT,
        columns=MOCK_COLUMNS,
    )
    result = ImageRepositoryManager().get_repository_row(MOCK_ROWS_DICT[0]["name"])
    assert result == MOCK_ROWS_DICT[0]


@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager._execute_schema_query"
)
@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager._conn"
)
def test_get_repository_row_no_repo_found(mock_execute, mock_connection, mock_cursor):
    mock_execute.return_value = mock_cursor(
        rows=[],
        columns=MOCK_COLUMNS,
    )

    with pytest.raises(ClickException) as expected_error:
        ImageRepositoryManager().get_repository_row("IMAGES")
    assert "does not exist in database" in expected_error.value.message


@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager._execute_schema_query"
)
def test_get_repository_row_more_than_one_repo(mock_execute, mock_cursor):
    mock_execute.return_value = mock_cursor(
        rows=MOCK_ROWS_DICT + MOCK_ROWS_DICT,
        columns=MOCK_COLUMNS,
    )
    with pytest.raises(ClickException) as expected_error:
        ImageRepositoryManager().get_repository_row("IMAGES")
    assert "Found more than one image repository" in expected_error.value.message


@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager.get_repository_row"
)
def test_get_repository_url(mock_get_row, mock_cursor):
    expected_row = MOCK_ROWS_DICT[0]
    mock_get_row.return_value = expected_row
    result = ImageRepositoryManager().get_repository_url(repo_name="IMAGES")
    mock_get_row.assert_called_once_with("IMAGES")
    assert isinstance(expected_row, Dict)
    assert "repository_url" in expected_row
    assert result == f"https://{expected_row['repository_url']}"


@mock.patch(
    "snowflake.cli.plugins.spcs.image_repository.manager.ImageRepositoryManager.get_repository_row"
)
def test_get_repository_url_no_scheme(mock_get_row, mock_cursor):
    expected_row = MOCK_ROWS_DICT[0]
    mock_get_row.return_value = expected_row
    result = ImageRepositoryManager().get_repository_url(
        repo_name="IMAGES", with_scheme=False
    )
    mock_get_row.assert_called_once_with("IMAGES")
    assert isinstance(expected_row, Dict)
    assert "repository_url" in expected_row
    assert result == expected_row["repository_url"]
