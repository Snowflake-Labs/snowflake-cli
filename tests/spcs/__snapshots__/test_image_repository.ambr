# serializer version: 1
# name: test_create_cli[False-False]
  '''
  +-----------------------------------------------------------+
  | key    | value                                            |
  |--------+--------------------------------------------------|
  | status | Image Repository TEST_REPO successfully created. |
  +-----------------------------------------------------------+
  
  '''
# ---
# name: test_create_cli[False-True]
  '''
  +-----------------------------------------------------------+
  | key    | value                                            |
  |--------+--------------------------------------------------|
  | status | Image Repository TEST_REPO successfully created. |
  +-----------------------------------------------------------+
  
  '''
# ---
# name: test_create_cli[True-False]
  '''
  +-----------------------------------------------------------+
  | key    | value                                            |
  |--------+--------------------------------------------------|
  | status | Image Repository TEST_REPO successfully created. |
  +-----------------------------------------------------------+
  
  '''
# ---
# name: test_create_cli_replace_and_if_not_exists
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Options '--if-not-exists' and '--replace' are incompatible.                  │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
